src/index.ts

```diff
@@ -744,26 +755,63 @@ class Datastore extends DatastoreRequest {
    *  The location prefix of an App Engine project ID.
    *  Often this value is 's~', but may also be 'e~', or other location prefixes
    *  currently unknown.
-   * @returns {string} base64 endocded urlsafe key.
+   * @param {function} callback The callback function.
+   * @param {?error} callback.err An error returned while making this request
+   * @param {string} callback.urlSafekey base64 endocded urlsafe key.
    *
    * @example
    * const {Datastore} = require('@google-cloud/datastore');
    * const datastore = new Datastore();
    * const key = datastore.key(['Company', 'Google']);
    *
-   * datastore.keyToLegacyUrlsafe(key) //ag9ncmFzcy1jbHVtcC00NzlyEwsSB0NvbXBhbnkiBkdvb2dsZQw
+   * datastore.keyToLegacyUrlsafe(key, (err, urlSafekey) => {
+   *   if (err) {
+   *     // Error handling omitted.
+   *   }
+   *   console.log(urlSafekey);
+   * })
    *
-   * @example
-   * <caption>Create a complete url safe key using location prefix </caption>
+   * //-
+   * // Create a complete url safe key using location prefix
+   * //-
    * const {Datastore} = require('@google-cloud/datastore');
    * const datastore = new Datastore();
    * const key = datastore.key(['Task', 123]);
    * const locationPrefix = 's~';
    *
-   * datastore.keyToLegacyUrlsafe(key, locationPrefix) //ahFzfmdyYXNzLWNsdW1wLTQ3OXIKCxIEVGFzaxh7DA
+   * datastore.keyToLegacyUrlsafe(key, locationPrefix, (err, urlSafekey) => {
+   *   if (err) {
+   *     // Error handling omitted.
+   *   }
+   *   console.log(urlSafekey);
+   *
+   * //-
+   * // If the callback is omitted, we'll return a Promise.
+   * //-
+   * const {Datastore} = require('@google-cloud/datastore');
+   * const datastore = new Datastore();
+   * const key = datastore.key(['Task', 123]);
+   * const locationPrefix = 's~';
+   *
+   * const urlSafekey = await datastore.keyToLegacyUrlsafe(key, locationPrefix);
+   * console.log(urlSafekey);
    */
-  keyToLegacyUrlsafe(key: entity.Key, locationPrefix?: string): string {
-    return urlSafeKey.legacyEncode(this.projectId, key, locationPrefix);
+  keyToLegacyUrlsafe(
+    key: entity.Key,
+    locationPrefixOrCallback?: string | KeyToLegacyUrlsafeCallback,
+    callback?: KeyToLegacyUrlsafeCallback
+  ): Promise<string> | void {
+    const locationPrefix =
+      typeof locationPrefixOrCallback === 'string'
+        ? locationPrefixOrCallback
+        : '';
+    callback =
+      typeof locationPrefixOrCallback === 'function'
+        ? locationPrefixOrCallback
+        : callback;
+    this.auth.getProjectId((err, projectId) => {
+      callback!(err, urlSafeKey.legacyEncode(projectId!, key, locationPrefix));
```

Comment:
When there's an error, we should not return the best-effort encoded key. That could lead users to believe it worked successfully.

Reply:
@laljikanjareeya 
something like
```ts
this.auth.getProjectId((err, projectId) => {
  if (err) {
    callback!(err);
    return;
  }
  callback!(null, urlSafeKey.legacyEncode(projectId!, key, locationPrefix));
});
```
and add a test, something along
```ts
    it('should not return urlSafeKey to user if auth.getProjectId erros', () => {
      const error = new Error('Some error');
      // tslint:disable-next-line: no-any
      (datastore.auth as any).getProjectId = (callback: Function) => {
        callback(error);
      };
      datastore.keyToLegacyUrlsafe({} as entity.Key, (err, urlSafekey) => {
        assert.strictEqual(err, error);
        assert.strictEqual(urlSafekey, undefined);
      });
    });

```